### DISCUSSION TITLE
`bud build` does not work

### DISCUSSION THREAD

**Discussion by pyrox0** [nrdxp/nixdev#417] [Q&A] [answered]
I'm trying to make a `devos`-inspired configuration(I've copied most of the code, but I had done a lot of work beforehand so I didn't want to start over. Most of the devOS code is in my repo.) I'm getting this error when running `bud build` to generate my config:
```
error: anonymous function at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/types.nix:533:38 called with unexpected argument 'priority'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/lists.nix:116:32:

          115|   */
          116|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          117|
(use '--show-trace' to show detailed location information)
```
<details>
<summary>With `--show-trace`</summary>
error: anonymous function at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/types.nix:533:38 called with unexpected argument 'priority'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/lists.nix:116:32:

          115|   */
          116|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          117|

       … while evaluating anonymous lambda

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/lists.nix:116:29:

          115|   */
          116|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                             ^
          117|

       … from call site

       … while evaluating 'isFunction'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/trivial.nix:403:16:

          402|   */
          403|   isFunction = f: builtins.isFunction f ||
             |                ^
          404|     (f ? __functor && isFunction (f.__functor f));

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:271:12:

          270|       loadModule = args: fallbackFile: fallbackKey: m:
          271|         if isFunction m || isAttrs m then
             |            ^
          272|           unifyModuleSyntax fallbackFile fallbackKey (applyIfFunction fallbackKey m args)

       … while evaluating 'loadModule'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:270:53:

          269|       # Like unifyModuleSyntax, but also imports paths and calls functions if necessary
          270|       loadModule = args: fallbackFile: fallbackKey: m:
             |                                                     ^
          271|         if isFunction m || isAttrs m then

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:311:22:

          310|           let
          311|             module = loadModule args parentFile "${parentKey}:anon-${toString n}" x;
             |                      ^
          312|             collectedImports = collectStructuredModules module._file module.key module.imports args;

       … while evaluating anonymous lambda

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:328:31:

          327|           disabledKeys = map moduleKey disabled;
          328|           keyFilter = filter (attrs: ! elem attrs.key disabledKeys);
             |                               ^
          329|         in map (attrs: attrs.module) (builtins.genericClosure {

       … from call site

       … while evaluating 'filterModules'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:324:36:

          323|       # modules recursively. It returns the final list of unique-by-key modules
          324|       filterModules = modulesPath: { disabled, modules }:
             |                                    ^
          325|         let

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:335:7:

          334|     in modulesPath: initialModules: args:
          335|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);
             |       ^
          336|

       … while evaluating anonymous lambda

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:334:37:

          333|
          334|     in modulesPath: initialModules: args:
             |                                     ^
          335|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:180:25:

          179|       merged =
          180|         let collected = collectModules
             |                         ^
          181|           (specialArgs.modulesPath or "")

       … while evaluating 'reverseList'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/lists.nix:393:17:

          392|   */
          393|   reverseList = xs:
             |                 ^
          394|     let l = length xs; in genList (n: elemAt xs (l - n - 1)) l;

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:184:33:

          183|           ({ inherit lib options config specialArgs; } // specialArgs);
          184|         in mergeModules prefix (reverseList collected);
             |                                 ^
          185|

       … while evaluating 'byName'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:445:25:

          444|       */
          445|       byName = attr: f: modules:
             |                         ^
          446|         zipAttrsWith (n: concatLists)

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:462:21:

          461|       # an attrset 'name' => list of submodules that declare ‘name’.
          462|       declsByName = byName "options" (module: option:
             |                     ^
          463|           [{ inherit (module) _file; options = option; }]

       … while evaluating the attribute 'matchedOptions'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:506:14:

          505|     in {
          506|       inherit matchedOptions;
             |              ^
          507|

       … while evaluating 'mapAttrsRecursiveCond'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/attrsets.nix:296:36:

          295|   */
          296|   mapAttrsRecursiveCond = cond: f: set:
             |                                    ^
          297|     let

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:192:28:

          191|           # For definitions that have an associated option
          192|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                            ^
          193|

       … while evaluating the attribute 'config'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:257:9:

          256|         options = checked options;
          257|         config = checked (removeAttrs config [ "_module" ]);
             |         ^
          258|         _module = checked (config._module);

       … while evaluating 'merge'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/types.nix:569:22:

          568|         check = x: isAttrs x || isFunction x || path.check x;
          569|         merge = loc: defs:
             |                      ^
          570|           (base.extendModules {

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:647:59:

          646|       if isDefined then
          647|         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal
             |                                                           ^
          648|         else let allInvalid = filter (def: ! type.check def.value) defsFinal;

       … while evaluating the attribute 'mergedValue'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:645:5:

          644|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          645|     mergedValue =
             |     ^
          646|       if isDefined then

       … while evaluating the option `system.build':

       … while evaluating the attribute 'value'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:611:9:

          610|     in warnDeprecation opt //
          611|       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          612|         inherit (res.defsFinal') highestPrio;

       … while evaluating anonymous lambda

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/modules.nix:192:72:

          191|           # For definitions that have an associated option
          192|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          193|

       … from call site

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/attrsets.nix:304:20:

          303|               then recurse (path ++ [name]) value
          304|               else f (path ++ [name]) value;
             |                    ^
          305|         in mapAttrs g;

       … while evaluating 'g'

       at /nix/store/gvnhn3wxc6pbzr67fa4bnqrk6lj78bj6-source/lib/attrsets.nix:301:19:

          300|           g =
          301|             name: value:
             |                   ^
          302|             if isAttrs value && cond value

       … from call site
</details>
My System:
OS: Arch(switching to nixOS when I get my config working :| )
Nix version: nix (Nix) 2.7.0pre20220124_0a70b37
My config repo: https://git.sr.ht/~mrhedgehog/nix-config/tree
LMK if you need any more details to help. Thanks for all the help you can give me!

**Comment by nrdxp** [comment, marked as answer]
So this is a bit tough to get to the bottom of from just what you have posted so far, because nothing in the trace comes from any of your local code from what I can see. In cases like this where I am desparate to debug an error that the trace isn't much help with, I either start from a minimal config and start adding sections of code until the error recurs, or if that's too much effort I do the opposite and start commenting out until the error goes away.

Wish their was something better to offer, but unless you have more context that may help us nail it down that might be your best bet. Nix definitely needs to improve its debug tooling :sweat_smile:

**Reply by pyrox0** [reply]
Alright, I'll do what I can. Thanks for the help!

**Reply by pyrox0** [reply]
After some testing, it's to do with my `hosts` block in my flake.nix. Hmmm.

**Reply by tgunnoe** [reply]
This is the nixpkgs commit now in nixos-unstable that breaks devos/digga.
https://github.com/NixOS/nixpkgs/commit/ccb85a53b6a496984073227fd8c4d4c58889f421

**Reply by tgunnoe** [reply]
https://github.com/divnix/devos/issues/419

**Reply by pyrox0** [reply]
I saw that issue. I figured that it was that commit. Do you have any idea what the solution could be?
